Rework Signup flow to include Login File download

sisou opened the pull request:
This PR updates the Keyguard Signup flow to the new UX and UI. It includes the following major features:

- New optional description added to DownloadLoginFile component
- New LoginfileAnimation component that animates a sketch into a stylized Login File based on steps
- New Overlay component, following the new overlay design already present in the Wallet, compatible with the FlippableHandler. The overlay works as a page, animating in when the `id` of the overlay becomes the page `:target` (through the URL fragment).
- Update Login File: replaced identicon circle with key-and-stars, removed ImageWallet logo
- Add mandatory Login File download to Signup flow, including an extra Login File explainer page

sisou committed:
Update identicon selection page layout

sisou committed:
Add Overlay styles, add address selection overlay, add LoginFile animation

sisou committed:
Add Login File download to create flow, update Login File visuals

sisou committed:
Add Login File explainer page

sisou committed:
Update button text to "Got it"

sisou committed:
Rework overlays

sisou committed:
Remove ImageWallet logo from Login File

sisou committed:
Add a waiting animation to the LoginfileAnimation

sisou committed:
Finalize LoginfileAnimation

sisou committed:
Cleanup

sisou committed:
Update translations

danimoh commented on src/components/LoginfileAnimation.css:
```diff
@@ -0,0 +1,153 @@
+
+.loginfile-animation .border:nth-child(1) { stroke-dasharray: 92 440 585 0; animation: border-1 3s ease-in-out infinite; }
+.loginfile-animation .border:nth-child(2) { stroke-dasharray: 0 102 50 585; animation: border-2 3s ease-in-out infinite; }
+.loginfile-animation .border:nth-child(3) { stroke-dasharray: 0 230 195 585; animation: border-3 3s ease-in-out infinite; }
```
When switching to the colored login file, the border border or animation is visible again for a short moment.
Would be good to avoid this.

danimoh commented on src/components/LoginfileAnimation.js:
```diff
@@ -0,0 +1,109 @@
+        for (let i = 8; i > step; i--) {
+            this.$el.classList.remove(`step-${i}`);
+        }
+        for (let i = 0; i <= Math.min(step, 8); i++) {
```
As the 8 correlates to the minimum password length, it could be a constant somewhere.

danimoh commented on src/components/LoginfileAnimation.css:
```diff
@@ -0,0 +1,153 @@
+.loginfile-animation {
```
Inconsistent file naming. Should be `LoginFileAnimation`.

danimoh commented on src/components/LoginfileAnimation.js:
```diff
@@ -0,0 +1,109 @@
+/* global TemplateTags */
+/* global LoginFileConfig */
+
+class LoginfileAnimation { // eslint-disable-line no-unused-vars
```
Inconsistent file and class naming. Should be `LoginFileAnimation`.

danimoh commented on src/components/LoginfileAnimation.js:
```diff
@@ -0,0 +1,109 @@
+     */
+    getElement() {
+        return this.$el;
+    }
```
Unused.

sisou replied:
This method is part of the standard API of Keyguard components.

danimoh commented on src/components/Overlay.css:
```diff
@@ -0,0 +1,135 @@
+     * [-1 *] so it animates from and to the top, and we add a little extra [1.1 = +10%] so the easing
+     * doesn't stop at the top of the viewport.
+     */
+    transform: translate3D(0, calc(-1.1 * ((100vh - 100%) / 2 + 100%)), 0);
```
The overlay can still be reached and interacted with via keyboard, causing an error later on.

sisou replied:
Fixed this in b0cc26009dadc2d09a01fd4d4ac8372908a2445c.

danimoh commented on src/components/Overlay.css:
```diff
@@ -0,0 +1,135 @@
+     * To adapt the animation distance to the screen height (because we don't want the overlay to only
+     * animate part of the way on larger screens, or ultra-fast on smaller screens), we calculcate the
+     * distance between the top of the overlay to the top of the viewport [(100vh - 100%) / 2] and add
+     * the heigh of the overlay to it [+ 100%]. Then we just turn it around into a negative distance
```
height

danimoh commented on src/components/Overlay.css:
```diff
@@ -0,0 +1,135 @@
+     */
+    transform: translate3D(0, calc(-1.1 * ((100vh - 100%) / 2 + 100%)), 0);
+    transition: transform var(--overlay-transition-time) cubic-bezier(0.3, 0, 0, 1),
+        opacity 0s var(--overlay-transition-time);
```
`0s` shouldn't be necessary.

sisou replied:
It helps with legibility/understanding and doesn't hurt.

danimoh commented on src/request/create/Create.js:
```diff
@@ -25,30 +31,57 @@ class Create {
+        const $loginfilePreviewImage = (document.getElementById('loginfile-preview'));
+
+        /** @type {HTMLButtonElement} */
+        const $loginfileExplainerBackButton = (document.getElementById('loginfile-explainer-go-back'));
```
Inconsistent capitalization of `loginFile`

danimoh commented on src/request/create/Create.js:
```diff
@@ -60,40 +93,79 @@ class Create {
+        });
+
+        this._passwordSetter.on(PasswordSetterBox.Events.SUBMIT, /** @param {string} password */ async password => {
+            // TODO: Save key to disk here?
```
I think doing it later is ok.

danimoh commented on src/request/create/Create.js:
```diff
@@ -60,40 +93,79 @@ class Create {
+            });
+            this._downloadLoginFile.on(DownloadLoginFile.Events.DOWNLOADED, () => {
+                this.finish(request);
+            });
```
If the user navigates back and forth, these event listeners and especially the `finish` one get assigned multiple times.
Could this result in the key being stored multiple times?

danimoh commented on src/components/Overlay.css:
```diff
@@ -0,0 +1,135 @@
+     * [-1 *] so it animates from and to the top, and we add a little extra [1.1 = +10%] so the easing
+     * doesn't stop at the top of the viewport.
+     */
+    transform: translate3D(0, calc(-1.1 * ((100vh - 100%) / 2 + 100%)), 0);
```
Sometimes a page scrolling is visible due to the translated overlay.

sisou replied:
I found that only happens if I refresh when the devtools are open for the popup. I think Chrome is then loading CSS in an undetermined order, sometimes even after the page content, leading to the glitch. Can you re-create the issue while devtools are closed?

danimoh replied:
I couldn't even reproduce it anymore with dev tools open.
I guess we just have to see whether it ever occurs again.

sisou committed:
Refactor LoginfileAnimation => LoginFileAnimation

also loginfileAnimation => loginFileAnimation and loginfile-animation => login-file-animation

sisou committed:
Prevent border animation during transition to colored state

sisou committed:
Make the number of steps of the LoginFileAnimation a constant

mraveux commented on src/request/create/Create.css:
```diff
@@ -1,53 +1,93 @@
+    flex-direction: column;
+}
+
+.page#login-file-download.loginfile-download-initiated .page-header .nq-h1:not(.confirm-download),
```
Inconsistent naming for login-file / loginfile, i guess.

sisou committed:
Prevent tabbing to buttons in hidden overlay

sisou committed:
Fix/remove comments

sisou committed:
Assign event handlers for DownloadLoginFile only once

sisou committed:
Refactor more loginfile => loginFile in variable names

mraveux commented on src/lib/LoginFile.js:
```diff
@@ -192,7 +182,7 @@ class LoginFile {
         const img = new Image();
         const loaded = new Promise(resolve => {
-            img.onload = () => resolve();
+            img.onload = () => resolve(true);
```
Is this necessary? Couldn't this be just `img.onload = resolve;`?

sisou replied:
I added it because newer Typescript (the version loaded in VSCode) complaints that the Promise is not typed as `void`. I also recently read that creating short arrow functions is better than assigning to browser-native methods, as you never know how their API changes and with what parameters the method will then be called with. So better to define the arguments to use yourself. The `true` really has no effect here, just to shut up TS.

mraveux commented on src/request/create/index.html:
```diff
@@ -75,45 +85,112 @@
+                <div class="page-body nq-card-body">
+                    <div class="download-loginfile"></div>
+                </div>
+                <a href="#login-file-explainer" class="nq-link skip">
```
Missing i18n?

sisou replied:
Yes, well spotted!

mraveux commented on src/request/remove-key/index.html:
```diff
@@ -323,6 +323,7 @@ <h1 data-i18n="export-file-heading" class="nq-h1">Access your Login File</h1>
                         <svg class="nq-icon"><use xlink:href="../../../node_modules/@nimiq/style/nimiq-style.icons.svg#nq-arrow-left"/></svg>
                     </a>
                     <h1 class="nq-h1" data-i18n="import-words-download-loginfile">Save your Login File</h1>
+                    <h1 class="nq-h1 confirm-download " data-i18n="download-loginfile-successful">Download successful?</h1>
```
```suggestion
                    <h1 class="nq-h1 confirm-download" data-i18n="download-loginfile-successful">Download successful?</h1>
```

mraveux reviewed (approved):
There doesn't seems to be opacity transitions on Firefox for the svg parts, or those are quite laggy. 

Otherwise, it looks nice. 👍  Only minor comments.

sisou committed:
Add missing i18n tag
